pages/book/integers.mdx

```diff
@@ -0,0 +1,107 @@
+    oneByte: Int as int8 = 0; // ranges from -128 to 127 (takes 8 bit = 1 byte)
+    twoBytes: Int as int16;   // ranges from -32,768 to 32,767 (takes 16 bit = 2 bytes)
+
+    init() {
+        // needs to be initialized in the init() because it doesn't have the default value
+        self.twoBytes = 55*55;
+    }
+}
+```
+
+Integer serialization is also available for the fields of [Structs](/book/defining-types#structs) and [Messages](/book/defining-types#structs), as well as in key/value types of [maps](/book/types#maps):
+
+```tact
+struct StSerialization {
+    martin: Int as int8;
+}
+
+message MsgSerialization {
+    seamus: Int as int8;
+    mcFly: map<Int as int8, Int as int8>;
+}
+```
+
+Motivation is very simple:
+* Storing $1000$ $257$-bit integers in state [costs](https://docs.ton.org/develop/smart-contracts/fees#how-to-calculate-fees) about $0.184$ TON per year.
+* Storing $1000$ $32$-bit integers only costs $0.023$ TON per year by comparison.
+
+### Serialization types
+
+Name             | Inclusive range                           | Space taken
+:--------------- | :---------------------------------------: | :------------------------:
+`uint8{:tact}`   | $0$ to $255$                              | 8 bit = 1 byte
+`uint16{:tact}`  | $0$ to $65\_535$                          | 16 bit = 2 bytes
+`uint32{:tact}`  | $0$ to $4\_294\_967\_295$                 | 32 bit = 4 bytes
+`uint64{:tact}`  | $0$ to $2^{64} - 1$                       | 64 bit = 8 bytes
+`uint128{:tact}` | $0$ to $2^{128} - 1$                      | 128 bit = 16 bytes
+`uint256{:tact}` | $0$ to $2^{256} - 1$                      | 256 bit = 32 bytes
+`int8{:tact}`    | $-128$ to $127$                           | 8 bit = 1 byte
+`int16{:tact}`   | $-32\_768$ to $32\_767$                   | 16 bit = 2 bytes
+`int32{:tact}`   | $-2\_147\_483\_648$ to $2\_147\_483\_647$ | 32 bit = 4 bytes
+`int64{:tact}`   | $-2^{63}$ to $2^{63} - 1$                 | 64 bit = 8 bytes
+`int128{:tact}`  | $-2^{127}$ to $2^{127} - 1$               | 128 bit = 16 bytes
+`int256{:tact}`  | $-2^{255}$ to $2^{255} - 1$               | 256 bit = 32 bytes
+`int257{:tact}`  | $-2^{256}$ to $2^{256} - 1$               | 257 bit = 32 bytes + 1 bit
+`coins{:tact}`   | $0$ to $2^{120} - 1$                      | 120 bit = 15 bytes
+
+<Callout>
+  Read more on serialization here: [Compatibility with FunC](/book/func##convert-serialization)
+</Callout>
+
+## Operations
+
+All runtime calculations with numbers are done at 257-bits, so [overflows](https://en.wikipedia.org/wiki/Integer_overflow) are quite rare. Nevertheless, if any math operation overflows, an exception will be thrown, and the transaction will fail. You could say that Tact's math is safe by default.
+
+Note, that there is no problem with mixing variables of [different state sizes](#serialization) in the same calculation. At runtime, they are all the same type no matter what — $257$-bit signed, so they all will fit.
```

Comment:
Would be nice to mention the problem of integer overflow for storage. For instance, I read the value from `oneByte`, multiply the (let say non-zero) value by `256`, so there is no run-time integer overflow. And then I try to store the result back in `oneByte`. What happens next?

Reply:
Failure in the action phase, right?

Reply:
Failure in the compute phase, but exit code 5 ("Integer out of expected range")

Reply:
We can add an explanation here and add the link to https://docs.tact-lang.org/language/ref/exit-codes. Btw, it looks like the section `5: Integer out of expected range (compute phase)` cannot be linked precisely, no anchor there.

Reply:
Good call — I'll add an explanation, but won't link to exit-codes yet, due to #101 (inability to link) and #106 (re-write into the Book)

Reply:
Created an issue so we won't forget it: https://github.com/tact-lang/tact-docs/issues/107